kernel/typing.ml

```diff
@@ -180,45 +230,75 @@ let rec pseudo_u sg (fail: int*term*term-> unit) (sigma:SS.t) : (int*term*term)
           let b' = mk_App (Subst.shift 1 a) (mk_DB dloc i 0) [] in
           pseudo_u sg fail sigma ((q+1,b,b')::lst)
 
-        | Const (_,c), Const (_,c') when name_eq c c' -> keepon ()
+        (* A definable symbol is only be convertible with closed terms *)
         | Const (l,cst), t when not (Signature.is_static sg l cst) ->
-          ( match unshift_reduce sg q t with None -> warn () | Some _ -> keepon ())
+          ( try sure_occur_check sg q (fun k -> k <= q) t; keepon ()
+            with VarSurelyOccurs -> warn () )
         | t, Const (l,cst) when not (Signature.is_static sg l cst) ->
-          ( match unshift_reduce sg q t with None -> warn () | Some _ -> keepon ())
+          ( try sure_occur_check sg q (fun k -> k <= q) t; keepon ()
+            with VarSurelyOccurs -> warn () )
 
+        (* X = Y :  map either X to Y or Y to X *)
         | DB (l1,x1,n1), DB (l2,x2,n2) when n1>=q && n2>=q ->
            let (n,t) = if n1<n2
                        then (n1,mk_DB l2 x2 (n2-q))
                        else (n2,mk_DB l1 x1 (n1-q)) in
            pseudo_u sg fail (SS.add sigma (n-q) t) lst
+
+        (* X = t :
+           1) make sure that t is possibly closed
+           2) if by chance t already is closed, map X to t  *)
         | DB (_,_,n), t when n>=q ->
           begin
-            let n' = n-q in
-            match unshift_reduce sg q t with
-            | None -> warn ()
-            | Some ut ->
-               let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
-               if Subst.occurs n' t' then warn ()
-               else pseudo_u sg fail (SS.add sigma n' t') lst
+            try
```

Comment:
Can't you put pnly on `try ... with` at the top of the function?

Reply:
I think it would break tail recursion of this function, no ?

Reply:
I don't understand this point, the block `try ... with` is already outside some recursive calls so for me it is already not tail rec. But my comment is maybe not accurate since it only trigger a warning